<a href="https://colab.research.google.com/github/nuramaliasusanti/nuramaliasusanti/blob/main/Mini_Project_Data_Cleansing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **DATA PREPROCESSING : DATA CLEANSING**

Dalam notebook ini akan dilakukan proses penyiapan data (data preprocessing) yakni data cleansing sebelum membuat model prediksi yang tepat untuk menentukan pelanggan akan berhenti (churn) atau tidak.

## Import Packages/Library

Melakukan import library/packages yang akan digunakan sebagai tools untuk mengeksekusi data. Adapun packages yang akan di import adalah pandas. Pandas (Python for Data Analysis) adalah library Python yang fokus untuk proses analisis data seperti manipulasi data, persiapan data, dan pembersihan data.

In [ ]:
# Import library
import pandas as pd

## Import Dataset

Melakukan import dataset ke dalam workspace disertai dengan menampilkan beberapa informasi dalam dataset

### Read CSV File

Melakukan pembacaan data dengan menggunakan dataframe. Penggunaan dataframe memudahkan dalam membaca data karena ditampilkan dalam schema yang jelas dan juga perubahan-perubahan yang akan dilakukan dapat diamati secara langsung agar sesuai dengan hasil yang diharapkan. 

In [ ]:
# Membaca file csv
df_clean = pd.read_csv('https://storage.googleapis.com/dqlab-dataset/dqlab_telco.csv')

In [ ]:
# Menampilkan jumlah kolom dan baris data
df_clean.shape 

(7113, 22)

In [ ]:
# Menampilkan gambaran isi data (5 baris teratas dataset)
df_clean.head() 

UpdatedAt   customerID  gender  ...  MonthlyCharges TotalCharges Churn
0     202006  45759018157  Female  ...           29.85        29.85    No
1     202006  45557574145    Male  ...           56.95      1889.50    No
2     202006  45366876421    Male  ...           53.85       108.15   Yes
3     202006  45779536532    Male  ...           42.30      1840.75    No
4     202006  45923787906  Female  ...           70.70       151.65   Yes

[5 rows x 22 columns]

In [ ]:
# Menampilkan jumlah ID Customer yang berbeda
df_clean['customerID'].nunique()

7017

Setelah menampilkan dataset, terlihat bahwa pada dataset yang akan digunakan terdapat data sebanyak 7113 baris dan 22 kolom dengan jumlah ID Customer yang berbeda sebanyak 7017. 
Hal ini mengakibatkan harus dilakukan filtering data untuk memastikan data yang diambil sudah benar karena pada saat proses load data belum diketahui kevalidan isi data. 

## Mencari Validitas ID Number Customer

Langkah selanjutnya adalah melakukan filtering terhadap data untuk memastikan kevalidan dari data yang digunakan. Adapun format ketentuannya adalah customer yang terdata harus memiliki ID Number yang valid.





### Filtering Data

Pada proses filtering data ini, perlu dilakukan dua hal, diantaranya:
1.   Memfilter ID Number Customer sesuai dengan kriteria
2.   Memfilter duplikasi ID Number Customer 



**1. Memfilter ID Number Customer sesuai Kriteria**


Kriteria ID Number (Phone Number) Customer yang valid :
*   Panjang karakter adalah 11 - 12 karakter
*   Terdiri dari angka saja dan tidak diperbolehkan ada karakter selain angka
*   Diawali dengan angka 45 pada 2 digit pertama

In [ ]:
# Menambahkan kolom bantuan 'valid_ID'
df_clean['valid_ID'] = df_clean['customerID'].astype(str).str.match(r'(45\d{9,10})') 
df_clean

UpdatedAt   customerID  gender  ...  TotalCharges Churn valid_ID
0        202006  45759018157  Female  ...         29.85    No     True
1        202006  45557574145    Male  ...       1889.50    No     True
2        202006  45366876421    Male  ...        108.15   Yes     True
3        202006  45779536532    Male  ...       1840.75    No     True
4        202006  45923787906  Female  ...        151.65   Yes     True
...         ...          ...     ...  ...           ...   ...      ...
7108     202006  45117219093    Male  ...       3686.05   NaN     True
7109     202006  45564915800  Female  ...       1115.15   NaN     True
7110     202006  45655964195    Male  ...       2537.00   NaN     True
7111     202006  45899251814    Male  ...       1393.60   NaN     True
7112     202006  45427332508  Female  ...         89.35   NaN     True

[7113 rows x 23 columns]

Penambahan kolom baru 'valid_ID' bertujuan untuk memeriksa apakah data pada baris tertentu memiliki ID Number yang valid (True) sesuai dengan kriteria yang ditentukan. 

Selanjutnya akan dilakukan filtering data dimana data yang ditampilkan hanya baris data yang memiliki valid_ID bernilai True. Setelah itu kolom 'valid_ID' bisa dihilangkan karena tidak dibutuhkan setelah filtering data selesai.

In [ ]:
# Memfilter data dan menghapus kolom 'valid_ID'
df_clean = df_clean[df_clean['valid_ID'] == True].drop('valid_ID', axis=1)
df_clean

UpdatedAt   customerID  gender  ...  MonthlyCharges TotalCharges Churn
0        202006  45759018157  Female  ...           29.85        29.85    No
1        202006  45557574145    Male  ...           56.95      1889.50    No
2        202006  45366876421    Male  ...           53.85       108.15   Yes
3        202006  45779536532    Male  ...           42.30      1840.75    No
4        202006  45923787906  Female  ...           70.70       151.65   Yes
...         ...          ...     ...  ...             ...          ...   ...
7108     202006  45117219093    Male  ...           69.70      3686.05   NaN
7109     202006  45564915800  Female  ...           95.50      1115.15   NaN
7110     202006  45655964195    Male  ...           98.65      2537.00   NaN
7111     202006  45899251814    Male  ...           61.65      1393.60   NaN
7112     202006  45427332508  Female  ...           89.35        89.35   NaN

[7006 rows x 22 columns]

Dari data diatas maka diketahui bahwa data tersebut hanya memiliki **7006 baris** data dengan **ID Number yang valid**. 

**2. Memfilter Duplikasi ID Number Customer**

Tahap ini berguna untuk memastikan bahwa tidak ada ID Number Customer yang duplikat. Adapun biasanya duplikasi ID Number ini tipenya :
*   Duplikasi dikarenakan inserting melebihi satu kali dengan nilai yang sama tiap kolomnya 
*   Duplikasi dikarenakan inserting beda periode pengambilan data




In [ ]:
# Drop duplicate rows
df_clean.drop_duplicates()

UpdatedAt   customerID  gender  ...  MonthlyCharges TotalCharges Churn
0        202006  45759018157  Female  ...           29.85        29.85    No
1        202006  45557574145    Male  ...           56.95      1889.50    No
2        202006  45366876421    Male  ...           53.85       108.15   Yes
3        202006  45779536532    Male  ...           42.30      1840.75    No
4        202006  45923787906  Female  ...           70.70       151.65   Yes
...         ...          ...     ...  ...             ...          ...   ...
7108     202006  45117219093    Male  ...           69.70      3686.05   NaN
7109     202006  45564915800  Female  ...           95.50      1115.15   NaN
7110     202006  45655964195    Male  ...           98.65      2537.00   NaN
7111     202006  45899251814    Male  ...           61.65      1393.60   NaN
7112     202006  45427332508  Female  ...           89.35        89.35   NaN

[6999 rows x 22 columns]

In [ ]:
# Drop duplicate ID Customer sorted by Periode
df_clean = df_clean.sort_values('UpdatedAt', ascending=False).drop_duplicates(['customerID'])
df_clean

UpdatedAt   customerID  gender  ...  MonthlyCharges TotalCharges Churn
0        202006  45759018157  Female  ...           29.85        29.85    No
4727     202006  45315483266    Male  ...           20.50      1198.80    No
4738     202006  45236961615    Male  ...          104.10       541.90   Yes
4737     202006  45929827382  Female  ...          115.50      8312.75    No
4736     202006  45305082233  Female  ...           81.25      4620.40    No
...         ...          ...     ...  ...             ...          ...   ...
2366     202006  45184632574  Female  ...           70.15        70.15   Yes
2367     202006  45023517923  Female  ...           85.55        85.55   Yes
2368     202006  45665022680  Female  ...          117.15      8529.50    No
2369     202006  45957014561    Male  ...           99.25      6549.45    No
7112     202006  45427332508  Female  ...           89.35        89.35   NaN

[6993 rows x 22 columns]

Setelah dilakukan filtering untuk pengecekan kevalidasian data yang digunakan, maka tersisa **6993 baris** data dengan **ID Number** yang **valid** dan **tidak ada duplikasi**.

## Mengatasi Missing Values

Selanjutnya akan dilakukan pengecekan missing values pada data dengan tools yang terdapat pada packages pandas. Pengecekan missing values ini bertujuan untuk mendeteksi apakah ada data yang tergolong churn atau tidak.

Untuk mengatasi missing values ini dilakukan 2 proses, yakni :
1. Mengatasi Missing Values dengan penghapusan baris
2. Mengatasi Missing Values dengan pengisian nilai tertentu

**1. Mengatasi Missing Values dengan penghapusan baris**

Akan dilakukan penghapusan baris dari data-data yang tidak terdeteksi apakah termasuk *churn* atau tidak, dalam hal ini data yang dibuang adalah data yang tidak termasuk ke dalam golongan apapun. Diasumsikan data modeller hanya bisa menerima data yang benar ada flag *churn* nya atau tidak.

In [ ]:
# Mengecek jumlah missing values pada kolom Churn
df_clean['Churn'].isnull().sum()

43

Setelah dilakukan pengecekan terhadap kolom Churn, terdapat 43 baris data dengan missing value sehingga perlu dilakukan drop/ penghapusan baris data. 

In [ ]:
# Menghapus data dengan missing values
df_clean.dropna(subset=['Churn'], inplace=True)

In [ ]:
# Menampilkan total baris data
df_clean.shape

(6950, 22)

Dari data diatas terlihat bahwa data yang telah dilakukan drop missing value nya berubah menjadi **6950 baris** data.